instrumentation/opentelemetry-api-0.4/src/main/java/io/opentelemetry/auto/instrumentation/opentelemetryapi/trace/UnshadedSpanContext.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.opentelemetry.auto.instrumentation.opentelemetryapi.trace;
+
+import static io.opentelemetry.auto.instrumentation.opentelemetryapi.trace.Bridging.toUnshaded;
+
+import unshaded.io.opentelemetry.trace.SpanContext;
+import unshaded.io.opentelemetry.trace.SpanId;
+import unshaded.io.opentelemetry.trace.TraceFlags;
+import unshaded.io.opentelemetry.trace.TraceId;
+import unshaded.io.opentelemetry.trace.TraceState;
+
+public class UnshadedSpanContext extends SpanContext {
+
+  private final io.opentelemetry.trace.SpanContext shadedSpanContext;
+
+  public UnshadedSpanContext(final io.opentelemetry.trace.SpanContext shadedSpanContext) {
+    this.shadedSpanContext = shadedSpanContext;
+  }
+
+  io.opentelemetry.trace.SpanContext getShadedSpanContext() {
+    return shadedSpanContext;
+  }
+
+  @Override
+  public TraceId getTraceId() {
+    return toUnshaded(shadedSpanContext.getTraceId());
```

Comment:
Agree with @inikem the optimization isn't clear, now we copy every time the method is accessed - is the idea that doing this lazy is good because the agent, so exporters and the like will commonly be accessing the shaded version and user apps probably won't access all of these unshaded copies? I think this makes sense but may still want to memoize these results, wouldn't bother making it volatile though.